src/S3.elm

```diff
@@ -191,51 +222,81 @@ uploadFileHttpTask :
     , bucket : String
     }
     -> Task Http.Error Response
-uploadFileHttpTask { url, file, parts, key, bucket } =
+uploadFileHttpTask record =
     Http.riskyTask
         { method = "POST"
-        , headers =
-            []
-        , url = url
-        , body =
-            Http.multipartBody
-                (List.map (\( a, b ) -> Http.stringPart a b) parts
-                    ++ [ Http.filePart "file" file ]
-                )
-        , resolver =
-            Http.bytesResolver
-                (\response ->
-                    case response of
-                        Http.BadUrl_ badUrl ->
-                            Err (Http.BadUrl badUrl)
-
-                        Http.Timeout_ ->
-                            Err Http.Timeout
-
-                        Http.NetworkError_ ->
-                            Err Http.NetworkError
-
-                        Http.BadStatus_ metadata _ ->
-                            Err (Http.BadStatus metadata.statusCode)
-
-                        Http.GoodStatus_ metadata _ ->
-                            Maybe.map2
-                                (\etag location ->
-                                    { etag = etag |> String.replace "\"" ""
-                                    , location = location
-                                    , bucket = bucket
-                                    , key = key
-                                    }
-                                )
-                                (Dict.get "etag" metadata.headers)
-                                (Dict.get "location" metadata.headers)
-                                |> Result.fromMaybe
-                                    (Http.BadBody "ETag or Location missing on response header")
-                )
+        , headers = []
+        , url = record.url
+        , body = fileBodyWithParts record.file record.parts
+        , resolver = Http.bytesResolver (expectedResponse record)
         , timeout = Nothing
         }
 
 
+uploadFileHttpRequest :
+    (Result Http.Error Response -> msg)
+    ->
+        { url : String
+        , file : File
+        , parts : List ( String, String )
+        , key : String
+        , bucket : String
+        , timeout : Maybe Float
+        , tracker : Maybe String
+        }
+    -> Cmd msg
+uploadFileHttpRequest toMsg record =
+    Http.riskyRequest
+        { method = "POST"
+        , headers = []
+        , url = record.url
+        , body = fileBodyWithParts record.file record.parts
+        , expect = Http.expectBytesResponse toMsg (expectedResponse record)
+        , timeout = record.timeout
+        , tracker = record.tracker
+        }
+
+
+fileBodyWithParts : File -> List ( String, String ) -> Http.Body
+fileBodyWithParts file =
+    Http.multipartBody
+        << List.foldr (\( k, v ) -> (::) (Http.stringPart k v))
```

Comment:
Since `map` uses `foldr` under the covers, it was inefficient to map a list and append and new list. This just starts with the file as the initial seed for the accumulator.